OpenTable Block

scruffian opened the pull request:
This PR adds an OpenTable block, providing block-editor native equivalent functionality of the existing OpenTable embed.

It supports embed code pasted from the OpenTable widget creator:
https://www.opentable.com/widget/reservation/preview?rid=1&lang=en-US

#### Changes proposed in this Pull Request:
* This PR adds a new block, `jetpack/opentable`.

![opentable](https://user-images.githubusercontent.com/275961/70922086-680c5300-201d-11ea-9d30-aa4bf6ab67aa.gif)

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?

New feature.

Internal reference: pb5gDS-er-p2

The goal would be to release it as a beta block first, in Jetpack 8.1.
Internal reference: p1576850286103100-slack-explorers

#### Testing instructions:

This is currently marked as a beta block, you will need to add `define( 'JETPACK_BETA_BLOCKS', true );` to your `wp-config.php` file.

* Open the Block Editor.
* Try pasting in an embed code from https://www.opentable.com/widget/reservation/preview?rid=1&lang=en-US
* Try editing the block settings in the sidebar

#### Proposed changelog entry for your changes:
* Block Editor: Added a OpenTable block.

scruffian committed:
Add OpenTable block

davemart-in commented:
Nice work getting this stubbed out!

Here's what I'm currently seeing:

![stubbed-out](https://user-images.githubusercontent.com/5634774/71106770-e38d1200-218d-11ea-94c5-161bf55913be.png)

In comparing the this current state of this PR to the design mockups we've still got quite a few things to clean up:

![open-table-v1](https://user-images.githubusercontent.com/5634774/71106847-00c1e080-218e-11ea-9695-09ae2e6f75fc.png)

Would it be helpful for me to go through and itemize the changes that still need to be made, or do you want to take another pass at cleaning things up first?

pablinos committed:
Removed unnecessary CSS after rebase/merge

pablinos committed:
Prevent the reastaurant picker getting too large in the sidebar

zinigor commented:
Because this is still in progress as of now, I'm going to move this to the next milestone.

glendaviesnz commented:
A couple of things I noticed when testing this:

1. OpenTable has country specific domain names for the embeds, eg. https://www.opentable.com.au/widget/reservation/preview?rid=122582&lang=en-US has `www.opentable.com.au`, so [this regex](https://github.com/Automattic/jetpack/pull/14220/files#diff-a7bdc34d18d10acfd4271ad3b9113a07R19) needs to be loosened up a bit

2. When embed script is pasted in, a fetch is fired off with the embed as the query string, which causes a 403 from the OpenTable API.

<img width="625" alt="script-403" src="https://user-images.githubusercontent.com/3629020/71786177-9950cf80-306d-11ea-8058-3980825ab9dc.png">

Adding a check for the embed code before [running the search](https://github.com/Automattic/jetpack/pull/14220/files#diff-57641f42122dc10b681306d55c09026cR32) would be one way of fixing this. 

```javascript
	useEffect( () => {
		if ( ! embedRegex.test( searchTerm ) ) {
			throttledSearchRestaurants( searchTerm );
		}
	}, [ searchTerm ] );
```

glendaviesnz commented on extensions/blocks/opentable/restaurant-picker.js:
```diff
@@ -0,0 +1,76 @@
+import useRestaurantSearch from './use-restaurant-search';
+
+const MAX_SUGGESTIONS = 20;
+const embedRegex = /<script type=\'text\/javascript\' src=\'\/\/www.opentable.com\/widget\/reservation\/loader\?[^']+\'><\/script>/;
```
It looks like OpenTable has country specific domain names for the embeds, eg. https://www.opentable.com.au/widget/reservation/preview?rid=122582&lang=en-US has www.opentable.com.au, so this regex needs to be loosened up a bit.

pablinos replied:
I did wonder about that. We might need to check how we're building the URL embed too

Copons replied:
+1
It also has `.co.uk` and `.it` (just two I've randomly checked).

The TLD also determines the `domain` parameter of the widget, which as far as I can see, it's just the TLD without dots.
E.g.
`.com` -> `domain=com`
`.co.uk` -> `domain=couk`

jeherve replied:
Done in de41f11b473d6e39906a27b7632e5d2a2c00c2de

glendaviesnz commented on extensions/blocks/opentable/use-restaurant-search.js:
```diff
@@ -0,0 +1,36 @@
+	] );
+
+	useEffect( () => {
+		throttledSearchRestaurants( searchTerm );
```
When embed script is pasted in, the fetch is fired off with the embed as the query string, which causes a 403 from the OpenTable API.

<img width="625" alt="script-403" src="https://user-images.githubusercontent.com/3629020/71786177-9950cf80-306d-11ea-8058-3980825ab9dc.png">

Checking for the embed code before running the search would be one way of fixing this. 

```javascript
	useEffect( () => {
		if ( ! embedRegex.test( searchTerm ) ) {
			throttledSearchRestaurants( searchTerm );
		}
	}, [ searchTerm ] );
```

pablinos replied:
Agreed! We should probably not fire off requests when we know that it's not a restaurant name

pablinos replied:
I've implemented a couple of checks to prevent this from firing so much. This particular error came from the `/` characters in the embed code. I'll URL encode the search term

Copons commented:
There are a few inconsistencies on the front end rendering, likely caused by some different CSS opinions between OpenTable and TwentyTwenty, that I think we should address.

<img width="304" alt="Screenshot 2020-01-06 at 13 38 04" src="https://user-images.githubusercontent.com/2070010/71821776-e7c09700-308a-11ea-84fd-ddebb26a8cd9.png"><img width="232" alt="Screenshot 2020-01-06 at 13 45 02" src="https://user-images.githubusercontent.com/2070010/71821780-e8f1c400-308a-11ea-8a13-4a72d688d890.png">

These should be a good start, but I'm not exactly sure how to enqueue in the "Jetpack way" front-end styles (`view` or `style` files) for dynamic blocks.

```scss
.ot-dtp-picker {
  box-sizing: content-box;

  .ot-title {
    margin: 4px auto 12px auto;
  }

  .ot-dtp-picker-selector-link {
    text-decoration: none;
  }

  input[type='submit'] {
    text-transform: none;
    padding: 0;
  }
}
```

Copons commented on extensions/blocks/opentable/attributes.js:
```diff
@@ -0,0 +1,80 @@
+const optionValues = options => options.map( option => option.value );
+
+export const languageOptions = [
+	{ value: 'en-US', label: __( 'English-US', 'jetpack' ) },
```
Do we need to `__()` these labels?
They indicate language names, and are already translated.

scruffian replied:
Good point, that doesn't make sense!

Copons commented on extensions/blocks/opentable/edit.js:
```diff
@@ -0,0 +1,293 @@
+	);
+};
+
+export default function OpentableEdit( { attributes, setAttributes, className, clientId } ) {
```
Capital T dangit! 😛 
(It's uncapitalized in other places too)

Copons commented on extensions/blocks/opentable/edit.js:
```diff
@@ -0,0 +1,293 @@
+			return;
+		}
+
+		const searchParams = new URLSearchParams( src.search );
```
I can't seem to find the JP browsers support list, but just in case [`URLSearchParams` is not compatible with IE11](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams#Browser_compatibility).

scruffian replied:
Ah shame. Good catch

scruffian replied:
Ah shame. Good catch

pablinos replied:
What's our policy on [using polyfills](https://www.npmjs.com/package/url-search-params-polyfill)?

pablinos replied:
Also for future reference, I found some info about browser support at PCYsg-2bz-p2. It's pretty confusing when it comes to IE, but it seems like IE 11 is supported for now because Microsoft still provides security updates etc for current Windows versions.

Copons replied:
> What's our policy on [using polyfills](https://www.npmjs.com/package/url-search-params-polyfill)?

I'm not sure, but I've found a polyfill for `URL` and `URLSearchParams` used somewhere else in JP, and just imported here as well for good measure: https://github.com/Automattic/jetpack/pull/14220/commits/20b594f0c6d68fb354725ccc7718162dc722e1fe

jeherve replied:
> IE 11 is supported for now

It should be, yes, although things are currently broken :( #14273 

In any case, using a polyfill does seem like a good idea indeed.

Copons commented on extensions/blocks/opentable/attributes.js:
```diff
@@ -0,0 +1,80 @@
+
+export const getStyleOptions = rid =>
+	compact( [
+		{ value: 'standard', label: __( 'Standard (224 x 301 pixels)', 'jetpack' ) },
```
I wonder if the sizes in pixel here are useful or confusing.
I can see their usefulness for the widget, that is supposed to be used in narrow sidebars, but I think it's less interesting when rendered in the post content.

pablinos replied:
I see your point. What do you think @scruffian, shall we get rid of them?

scruffian replied:
I think we should just leave it so it aligns with their widget, and iterate later

pablinos replied:
👍

Copons commented on extensions/blocks/opentable/edit.js:
```diff
@@ -0,0 +1,293 @@
+					isCollapsed={ true }
+					icon="edit"
+					label={ __( 'Type', 'jetpck' ) }
+					controls={ styleOptions.map( styleOption => ( {
```
This feels a bit out of place to me.
As far as I know, no other block with style options let the user change it from the toolbar.
I'm not adverse to adding more functionality to the toolbar instead of the sidebar, but in this particular case, the pencil icon is (or maybe _was_? 🤔) used to toggle between placeholder and preview (see for example the Image block, with the relatively new swap/edit icon, which was a pencil once).

I guess we could do a few things here:
- Change the `edit` icon into something more style-related ([`admin-appearance`](https://developer.wordpress.org/resource/dashicons/#admin-appearance)?).
- Add a ["swap"](https://github.com/WordPress/gutenberg/blob/2c1e3be7d1cec76176db6cccee1c6add93cc3e4d/packages/block-library/src/image/icon.js#L8) button that reverts the block back to the placeholder state with the empty restaurant picker field.
- Remove the Embed Code section from the sidebar (its purpose would be resolved by the "swap" button).

scruffian replied:
I think this is how real style options work - they appear in both places - try it with an image block...

Copons replied:
@scruffian I don't see the Image styles in the toolbar, but only in the sidebar... 😮

scruffian replied:
Its hard to spot:
<img width="562" alt="Screenshot 2020-01-06 at 17 36 58" src="https://user-images.githubusercontent.com/275961/71836120-3aaa4680-30ab-11ea-989b-29f06efffcb3.png">

Copons replied:
TIL!
I had no idea that Core was overloading the transform button with block styles too!

This said, it's still slightly different than the OpenTable block.
I think that UI comes for free when you use the "vanilla block styles approach" (I've never coded them, so I'm totally unfamiliar with how they work), which looks slightly different than what we have done here.
I'll see if there's a way to simplify and make it more consistent with Core Guten.

scruffian replied:
Yeah, I share your concern. Happy to be lead by @lessbloat...

davemart-in replied:
Yeah, a couple of things:

- I think we're trying to move away from the situation where clicking the pencil takes you back to the blank slate state.
- Typically with style variants, we do show them both in the right sidebar, and in the transform dropdown, but these are not true style variants. We'd need to submit a patch in core to make these show up in the transform menu. I'm not against doing that, it will just mean that this will take longer to ship. If we decide to go that route, I'd prefer to do this in three phases: 1) Ship what we've got, 2) Make the patch in core, 3) Circle back and update this block.
- The pencil icon is a bit of a hack, but I like having it there. Otherwise on smaller resolutions, it's not clear that you can change the style of this embed.

**Our options**

A) Keep this as is. Maintain both the style options on the right and the style change option via the pencil icon dropdown in the toolbar.

B) Work through the 3 phases that I outlined above to make it possible to show these style variations in the transform menu. I'm okay proceeding with this option. We have no guarantee that it will be approved in core though.

C) Drop the pencil icon dropdown. Just show the style change in the right sidebar. My worry with this option is discoverability at lower screen resolutions.

D) Remove the pencil icon dropdown, then convert the current right sidebar style selector to a simple dropdown. My worry with this option is discoverability. I think many people will just gloss over the dropdown and not know that it's there.

Copons commented on extensions/blocks/opentable/edit.js:
```diff
@@ -0,0 +1,293 @@
+
+	const restaurantPicker = ( pickerProps = {} ) => (
+		<RestaurantPicker rids={ rid } { ...pickerProps } />
+	);
```
Is there a reason why we are not directly rendering this as:
```jsx
<RestaurantPicker rids={ rid } onChange={ onPickerSubmit } />

// and...

<RestaurantPicker
  rids={ rid }
  label={ __( 'Enter your restaurant name, OpenTable Restaurant ID or embed code', 'jetpack' ) }
  onSubmit={ onPickerSubmit }
/>
```
Looks simple enough to me! 🤔

scruffian replied:
🚢

Copons commented on extensions/blocks/opentable/restaurant-picker.js:
```diff
@@ -0,0 +1,76 @@
+			{ ...props }
+			onChange={ onChange }
+		/>
+	);
```
This looks a bit off to me 😮 

<img width="414" alt="Screenshot 2020-01-06 at 15 00 58" src="https://user-images.githubusercontent.com/2070010/71826244-66223680-3095-11ea-9629-869d83b11ade.png">

Is it supposed to use a serif font (on TwentyTwenty)?

Copons commented:
I'm wondering about the UX of the restaurant picker with the `FormTokenField` which only displays the ID.
A few months ago, I've had a similar issue on an old (now removed) FSE feature: it was a post selector for a given CPT with an autocomplete popover list, which displayed the post title, and stored its ID.

If I'm not mistaken, this is the last version of that component before its removal: https://github.com/Automattic/wp-calypso/blob/3932cbd44a2a26897d233d094c9f44ad231c8ba8/apps/full-site-editing/full-site-editing-plugin/full-site-editing/components/post-autocomplete/index.js

What do y'all think if we adapted it (and modernized it with React hooks as well!) to use the OpenTable API?

scruffian commented:
> What do y'all think if we adapted it (and modernized it with React hooks as well!) to use the OpenTable API?

Sounds like a good idea!

Copons committed:
Simplify `RestaurantPicker` rendering

Copons committed:
Don't translate already translated labels

pablinos commented:
I was wondering about the UX of displaying just the ID, and I ran up against some problems.

Using the `FormTokenField`, I thought we could have it display the restaurant name and the ID, but the problem is that it gets a bit long sometimes.

Also, when I looked into doing that, I couldn't find a way of retrieving specific restaurant details by ID from the API endpoint we're using. There are a few problems with that, but one of the biggest is if someone enters their ID and it's a subset of another, like 432 and 4321. You nearly always get back restaurants with 432 in the name, street address, or with IDs starting and ending in 432, but never the restaurant with ID 432.

The only official OpenTable API I could find didn't seem to offer this information either, but if it did, we'd have to proxy the requests and I'm not sure how that would work with access tokens etc. It looks like you have to become [a partner to use their Consumer API](https://dev.opentable.com/). We might want to go that route depending on what's involved.

As a side thought, would it be worth adding the features of your post autocomplete component to the `FormTokenField`, or maybe the whole thing as a new component to @wordpress/components? It seems like this would be generally useful.

Copons committed:
Import a polyfill to use URLSearchParams on IE11

Copons committed:
capital_T_dangit!

huguesvincent commented:
For the record, I filled this form back in December and approached a few folks at Open Table without any success so far. I'll keep trying to get on that whitelist.

scruffian committed:
change the edit icon to admin-appearance

Copons committed:
Add front end styles

pablinos committed:
Fixed CSS for mobile and TwentyTwenty theme

Copons committed:
Fix placeholder style affecting the token form

Copons commented:
Ah, I forgot about my `PostAutocomplete` component, it's not a good fit for this as it expects 1 selected results, while OpenTable allows for multiple.

> Also, when I looked into doing that, I couldn't find a way of retrieving specific restaurant details by ID from the API endpoint we're using.

@pablinos Right, damn!
We could partly solve this by storing the restaurant name as a block attribute (not super elegant, but at least I don't expect folks to save dozens of restaurants in their OT blocks?), but that wouldn't help when pasting the embed code, which doesn't contain the restaurant names at all.

> if someone enters their ID and it's a subset of another, like 432 and 4321. You nearly always get back restaurants with 432 in the name, street address, or with IDs starting and ending in 432, but never the restaurant with ID 432.

This is rather annoying.
We could be more transparent, and stop saying that the restaurant ID is an accepted value:
Something like this:
```diff
- Enter your restaurant name, OpenTable Restaurant ID or embed code
+ Enter your restaurant name, or paste an OpenTable Reservation Widget embed code
```

pablinos committed:
Merged suggestion list styles

pablinos commented on extensions/blocks/opentable/editor.scss:
```diff
@@ -66,20 +73,32 @@
+				margin: 4px 0 0 0;
+				position: absolute;
+				right: 0;
+				top: 4.4em;
```
@scruffian I don't feel good about this, but it works. Have you got any thoughts? The idea is to position the button to stop it jumping around with the suggestion list, but it's not part of the `FormTokenField`s container. 

On mobile, I've made it centred beneath the field, which also isn't fantastic. We could potentially use `calc` to change the width of the input, but either way, it's all a bit hacky. 

Ideas are very welcome!

scruffian replied:
I updated this to use flexbox which seems a bit more robust

pablinos committed:
Prevent API calls for blank queries or the embed code

pablinos commented:
> We could be more transparent, and stop saying that the restaurant ID is an accepted value

Yeah that could work @Copons, but then we'll display the ID, and we're also still stuck when they paste in the embed. I think we may have to leave it with the ID for now and see if we can get a better API endpoint to use.

pablinos committed:
Fixed the encoding of the boolean attributes in the URL

The widget was never rendering in an iframe, and it was because it
wasn't being set to the string 'true' in the query parameter.

kraftbj commented:
I haven't looked at anything else, but Travis is failing the JS linting due to:

```
/home/travis/build/Automattic/jetpack/extensions/blocks/opentable/icon.js

  4:1  error  Missing external, internal dependencies docblocks  wpcalypso/import-docblock
```

jeherve commented on extensions/blocks/opentable/edit.js:
```diff
@@ -0,0 +1,294 @@
+			/>
+			<div className={ `${ className }-placeholder-links` }>
+				<ExternalLink
+					href="https://en.support.wordpress.com/widgets/open-table-widget/"
```
Let's get block support URLs in here maybe, with a different link depending on your site type like it's done here?
https://github.com/Automattic/jetpack/blob/54d177da89d95b24f77f8886bf8061ad7cb51661/extensions/blocks/simple-payments/index.js#L28-L31

pablinos replied:
I'm not sure if these exist yet. I'll do some digging.

jeherve commented on extensions/blocks/opentable/edit.js:
```diff
@@ -0,0 +1,294 @@
+			<div className={ `${ className }-placeholder-links` }>
+				<ExternalLink
+					href="https://en.support.wordpress.com/widgets/open-table-widget/"
+					target="_blank"
```
I don't know if you need the `target` attribute here, since the component is already `ExternalLink`? I would assume it already handles that for you, but I have not tested.

pablinos replied:
You're right, it's already on the link

jeherve commented on extensions/blocks/opentable/edit.js:
```diff
@@ -0,0 +1,294 @@
+/**
+ * External dependencies
+ */
+import 'url-polyfill';
```
I think you'll need to specify a source here.

pablinos replied:
I'm not 100% sure what you mean here. This is how it's being [imported elsewhere](https://github.com/Automattic/jetpack/blob/7ee491958b6652baa38bcad791f2a36b151e504d/modules/search/instant-search/lib/query-string.js#L4) not that this means it's the right way to do it of course!

pablinos replied:
I read somewhere that all extensions receive all the polyfills by default, so maybe this line isn't needed at all?

jeherve replied:
> I read somewhere that all extensions receive all the polyfills by default,

It's only the WordPress polyfills though, as far as I know:
https://developer.wordpress.org/block-editor/contributors/develop/scripts/#polyfill-scripts

> This is how it's being imported elsewhere not that this means it's the right way to do it of course!

True, my bad! So I take it I don't need to specify the source like so?

```suggestion
import 'url-polyfill' from 'url-polyfill';
```

jeherve commented on extensions/blocks/opentable/edit.js:
```diff
@@ -0,0 +1,294 @@
+			<>
+				<div className={ `${ className }-overlay` }></div>
+				<iframe
+					title={ `Open Table Preview ${ clientId }` }
```
Is this something we'd want to translate?

scruffian replied:
✅

jeherve commented on extensions/blocks/opentable/edit.js:
```diff
@@ -0,0 +1,294 @@
+				<Toolbar
+					isCollapsed={ true }
+					icon="admin-appearance"
+					label={ __( 'Type', 'jetpck' ) }
```
```suggestion
					label={ __( 'Type', 'jetpack' ) }
```

jeherve commented on extensions/blocks/opentable/edit.js:
```diff
@@ -0,0 +1,294 @@
+					options={ languageOptions }
+				/>
+				<ToggleControl
+					label={ __( 'Load the widget in an iFrame (Recommended)', 'jetpack' ) }
```
Seeing that option, I wonder why someone would want to use a different option. Maybe we could make the choice for the site owner and choose for them on this?

pablinos replied:
This seems reasonable to me. Generally speaking not having an iframe is likely to cause trouble. @scruffian What do you think?

scruffian replied:
Makes sense to me

pablinos replied:
I decided to keep the option, just in case, and move it to the Advanced section.

jeherve commented on extensions/blocks/opentable/edit.js:
```diff
@@ -0,0 +1,294 @@
+			</PanelBody>
+			<PanelBody title={ __( 'Settings', 'jetpack' ) }>
+				<RestaurantPicker rids={ rid } onChange={ onPickerSubmit } />
+				<SelectControl
```
Instead of offering that option, I wonder if it wouldn't be best to follow the site language settings; if the site language is a language supported by OpenTable use that, if not default to English?

pablinos replied:
Good idea. Do you think we should have both? Default to the site language, but allow the user to override it?

jeherve replied:
That would work too. I am not sure how many folks would need to override it. Maybe folks with multilanguage sites? I'll let you all decide what seems to be the best user experience here!

scruffian replied:
Maybe we should punt this to V2, we dont know how popular this block will be yet...

pablinos replied:
You may have seen that I've implemented this by having an inline script that sets the site locale in the `core/block-editor' store. I'm not sure if it's the most elegant of solutions, so opinions are welcome!

jeherve commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -0,0 +1,155 @@
+		$url .= '&rid=' . wp_kses( $rid, array() );
+	}
+
+	return $url;
```
What do you think about adding a filter here, allowing one to change the final URL and add extra parameters such as the marketing tracking options offered in the widget builder?

scruffian replied:
✅

jeherve reviewed (changes requested):
Here is a first review, with inline comments below. 

As @kraftbj mentioned there is a small issue with unit tests failing right now; I think it may be caused by the url polyfill import missing a source.

CodeClimate also had some recommendations. Most of them can be ignored (I just did that), but 2 of them point towards the complexity of 2 functions, and it may be worth taking a look at it to see if things could be extracted to be made simpler maybe.

jeherve commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -0,0 +1,155 @@
+		return has_any_blog_stickers( array( 'premium-plan', 'business-plan', 'ecommerce-plan' ), $site_id );
+	}
+	// For all Jetpack sites.
+	return Jetpack::is_active() && Jetpack_Plan::supports( 'opentable' );
```
Since the block doesn't rely on the WordPress.com API to work, I am not sure we can pay gate its access, as per rule 5 of the guidelines:

https://developer.wordpress.org/plugins/wordpress-org/detailed-plugin-guidelines/#5-trialware-is-not-permitted

I believe it should remain free.

pablinos replied:
I think we've reached a decision on this one!

jeherve commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -0,0 +1,155 @@
+	// The OpenTable script uses multiple `rid` paramters,
+	// so we can't use WordPress to output it, as WordPress attempts to validate it and removes them.
+	// phpcs:ignore WordPress.WP.EnqueuedResources.NonEnqueuedScript
+	return '<script type="text/javascript" src="' . esc_url( jetpack_opentable_build_embed_url( $attributes ) ) . '"></script>';
```
Should we use `Jetpack_Gutenberg::block_classes()` here and wrap the block into a container using those classes? It would allow the classes that can be entered here to be used:

![image](https://user-images.githubusercontent.com/426388/71984067-69016080-3228-11ea-9654-22e2527f0bc5.png)

scruffian replied:
Good idea. Done ✅

jeherve reviewed (changes requested):
There seems to be an aligment issue with the main button (this is using Chrome):

![image](https://user-images.githubusercontent.com/426388/71982713-710bd100-3225-11ea-8c31-aa986c95ba44.png)

*************

> Separate with commas or the Enter key.

This sentence feels odd. The main input field does not mention that you can add several names / codes, so it is a bit confusing to see that sentence.

*******

![image](https://user-images.githubusercontent.com/426388/71983080-4c642900-3226-11ea-929b-7d0cb6067285.png)

It seems a bit odd to see scrollbars within the preview images here.

********

It would be nice if we could show the restaurant names here instead of the IDs, which may not mean much to most people:

![image](https://user-images.githubusercontent.com/426388/71983295-c09ecc80-3226-11ea-95a6-f24f644ebe77.png)

*****

When changing the block type, it does not get reflected (selected/highlighted) here:

![image](https://user-images.githubusercontent.com/426388/71983465-24c19080-3227-11ea-920f-85356df3e04e.png)

******

This is a bit different from the way we've typically done things in other blocks:

![image](https://user-images.githubusercontent.com/426388/71983835-f1cbcc80-3227-11ea-9e8d-0f715d66590e.png)

I didn't expect to see the option to edit my embed there, as I've been used to the pencil icon within the block itself, like in the Pinterest block:

![image](https://user-images.githubusercontent.com/426388/71983879-0ad47d80-3228-11ea-9aa7-8c03e6859921.png)
![image](https://user-images.githubusercontent.com/426388/71983895-14f67c00-3228-11ea-8bb7-e8e975eef738.png)

scruffian committed:
remove the embed code form from the sidebar

scruffian commented on extensions/blocks/opentable/icon.js:
```diff
@@ -0,0 +1,15 @@
+	<SVG height="18" viewBox="0 0 25 18" width="25" xmlns="http://www.w3.org/2000/svg">
+		<Path
+			d="m123.5 77.2858191c0-1.2028985.974684-2.1783159 2.176848-2.1783159 1.202165 0 2.177215.9754174 2.177215 2.1783159 0 1.2032655-.97505 2.1786828-2.177215 2.1786828-1.202164 0-2.176848-.9754173-2.176848-2.1786828zm15.29169 2.1786828c-1.202349 0-2.176849-.9754173-2.176849-2.1786828 0-1.2028985.9745-2.1783159 2.176849-2.1783159 1.202164 0 2.177215.9754174 2.177215 2.1783159 0 1.2032655-.975051 2. ...
+			fill="#c93642"
```
@lessbloat what do you think?

davemart-in replied:
This has come up a couple of times. Many of the core embed blocks are colored, which is why we went colored with this block. Here's [a monochrome version](https://p111.p2.n0.cdn.getcloudapp.com/items/llu4gY5b/OpenTable-logo-mono.svg). Happy to swap it over.

jeherve replied:
I don't think that's a big issue, but since all other Jetpack block icons are monochrome at the moment, I think it may make sense to keep following that rule so no icon stands out when looking at all blocks in the Jetpack category?

scruffian committed:
fix the placeholder styles

scruffian committed:
fix the restaurant picker in the sidebar

scruffian committed:
fix dependencies

scruffian committed:
add a wrapper so that we can use advanced class names

scruffian commented on extensions/blocks/opentable/index.js:
```diff
@@ -0,0 +1,52 @@
+	keywords: [ 'opentable', 'reservation', 'restaurant' ],
+
+	supports: {
+		html: false,
```
I think that might be something for v2. Thoughts @lessbloat?

davemart-in replied:
Good call. Happy to include alignment. If we can squeeze it in for the v1 launch, let's go for it. If that risks this block not shipping in the next JP release, let's hold off and add it later.

jeherve replied:
Tracked here:
https://github.com/Automattic/jetpack/issues/14335

scruffian committed:
add an oxford comma

scruffian commented on extensions/blocks/opentable/edit.js:
```diff
@@ -0,0 +1,294 @@
+			<RestaurantPicker
+				rids={ rid }
+				label={ __(
+					'Enter your restaurant name, OpenTable Restaurant ID or embed code',
```
✅

scruffian committed:
Update extensions/blocks/opentable/edit.js

Co-Authored-By: Jeremy Herve <jeremy@jeremy.hu>

scruffian committed:
translate the iframe title

scruffian committed:
make the keywords translatable

scruffian commented on extensions/blocks/opentable/index.js:
```diff
@@ -0,0 +1,52 @@
+
+	category: 'jetpack',
+
+	keywords: [ 'opentable', 'reservation', 'restaurant' ],
```
✅

scruffian commented on extensions/blocks/opentable/index.js:
```diff
@@ -0,0 +1,52 @@
+
+export const settings = {
+	title,
+
```
✅

scruffian committed:
update the release number

scruffian commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -0,0 +1,155 @@
+/**
+ * OpenTable Block.
+ *
+ * @since 8.x
```
✅

scruffian committed:
add a check to see if the rid attribute is set

scruffian commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -0,0 +1,155 @@
+ * @return string The filtered attribute
+ */
+function jetpack_opentable_block_get_type_attribute( $attributes ) {
+	if ( count( $attributes['rid'] ) > 1 ) {
```
✅

scruffian committed:
add a filter to the URL

scruffian committed:
make it clear you can add more than one restaurant

scruffian committed:
dont show scrollbars in the preview

pablinos committed:
Check that the attribute name exists in the defaults

It was brought up that the attribute name may not be valid/exist, so
this adds a check for that.

pablinos committed:
Added some further documentation about how the theme attribute works

scruffian committed:
bold the selected style option

scruffian commented:
> When changing the block type, it does not get reflected (selected/highlighted) here:

I added a commit to fix this, but it will only work with Gutenberg 6.0+

pablinos committed:
Updated the escaping of the embed URL

It now escapes the final URL rather than the individual query
parameters.

pablinos committed:
Added further explanation to the style/theme parameters

pablinos committed:
Added even more explanation on the confusing theme and type parameters

I think I was confused myself! This seems like a good enough explanation
for now though

pablinos committed:
Added validation to the possible theme values

pablinos committed:
Namespaced the CSS so it wouldn't accidentally clash with a theme

pablinos committed:
Refactored to a namespace.

This is so we can have some constants and functions for repeated values
and operations.

pablinos committed:
Refactored repeated string to a constant and added is_wpcom function

pablinos committed:
Fixed the URL filter name

pablinos committed:
Changed to use `get_current_blog_id`

pablinos committed:
Refactored the block registration and availability to hooks

pablinos committed:
Removed target property from the external links

pablinos committed:
Added a setting so the default locale is the same as the site's

pablinos committed:
Removed availability logic to make it available by default

pablinos committed:
Added overflow: hidden to the style preview, to remove scrollbars

pablinos commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -0,0 +1,155 @@
+ *
+ * @return string The filtered attribute
+ */
+function jetpack_opentable_block_get_attribute( $attributes, $attribute_name ) {
```
I've added some validation code in the function. If someone calls this a null value, they'll rightly get an error

pablinos commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -0,0 +1,155 @@
+}
+
+/**
+ * Get the block theme attribute
```
I've had a stab at explaining this in a fairly lengthy comment. It's generally a bit confusing. There are two parameters, where one will do, so we're mapping between our `style` and their `theme` and `type`.

pablinos commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -0,0 +1,155 @@
+ *
+ * @return string Embed URL
+ */
+function jetpack_opentable_build_embed_url( $attributes ) {
```
A lot of this escaping has been consolidated into a single call to `esc_url`. I'd be interested to get your opinion on that.

pablinos commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -0,0 +1,155 @@
+		if ( in_array( $attribute_name, array( 'iframe', 'newtab' ), true ) ) {
+			return $attributes[ $attribute_name ] ? 'true' : 'false';
+		}
+		return wp_kses( $attributes[ $attribute_name ], array() );
```
As mentioned above, we've moved to using `esc_url`

pablinos commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -0,0 +1,155 @@
+		'missing_plan',
+		array(
+			'required_feature' => 'opentable',
+			'required_plan'    => ( defined( 'IS_WPCOM' ) && IS_WPCOM ) ? 'value_bundle' : 'jetpack_premium',
```
I've implemented this, but we've also removed a lot of this code

pablinos committed:
Improved the embed code regex to cope with other TLDs

pablinos committed:
Encoded the restaurant serach term for added safety

pablinos committed:
Moved the iframe option to the advanced inspector controls

jeherve commented on extensions/blocks/opentable/opentable.php:
```diff
@@ -0,0 +1,169 @@
+		}
+	}
+
+	return apply_filters( 'jetpack_opentable_block_url', esc_url( $url ) );
```
Since we're `esc_url` at the very end, as we add the URL here, maybe we don't need it here anymore?
https://github.com/Automattic/jetpack/blob/add/opentable/extensions/blocks/opentable/opentable.php#L53

It's also a good opportunity to add a docblock for that filter, so our parser can generate a page for that filter in the Jetpack Codex.

I would also recommend passing the attributes to the filter, since one could use them when modifying the final URL.

```suggestion
	/**
	 * Filter the OpenTable URL used to embed a widget.
	 *
	 * @since 8.2.0
	 *
	 * @param string $url        OpenTable embed URL.
	 * @param array  $attributes Array of block attributes.
	 */
	return apply_filters( 'jetpack_opentable_block_url', $url, $attributes );
```

pablinos replied:
That's much better, thanks!

jeherve commented on extensions/blocks/opentable/restaurant-picker.js:
```diff
@@ -0,0 +1,76 @@
+			suggestions={ restaurantNames }
+			onInputChange={ setInput }
+			maxSuggestions={ MAX_SUGGESTIONS }
+			label={ _n( 'Restaurant', 'Restaurants', selectedRestaurants.length ) }
```
I think we'll need a textdomain here for that translatable string.

pablinos replied:
Nice spot, thanks!

jeherve commented on class.jetpack-plan.php:
```diff
@@ -145,6 +145,7 @@ public static function get() {
 
 		if ( in_array( $plan['product_slug'], $premium_plans, true ) ) {
 			$supports[]    = 'akismet';
+			$supports[]    = 'opentable';
```
I think we can remove this now?

```suggestion
```

jeherve commented on class.jetpack-plan.php:
```diff
@@ -167,6 +168,7 @@ public static function get() {
 
 		if ( in_array( $plan['product_slug'], $business_plans, true ) ) {
 			$supports[]    = 'akismet';
+			$supports[]    = 'opentable';
```
```suggestion
```

pablinos replied:
I've been meaning to do that, thanks for spotting it for me!

pablinos committed:
Remove unnecessary addition to the business plan features

Co-Authored-By: Jeremy Herve <jeremy@jeremy.hu>

pablinos committed:
Remove unnecessary addition to the plan features

Co-Authored-By: Jeremy Herve <jeremy@jeremy.hu>

pablinos committed:
Added a docblock to the URL filter

Co-Authored-By: Jeremy Herve <jeremy@jeremy.hu>

pablinos committed:
[not verified] Fixed placeholder links padding of new layout

pablinos committed:
[not verified] Updated OpenTable sign up URL

pablinos committed:
[not verified] Explicitly make the embed button secondary

pablinos committed:
[not verified] Fixed CSS for the embed button to mimic the latest Gutenberg

jeherve committed:
Add missing textdomains

jeherve committed:
Register block without hooking into init

jeherve committed:
Update to monochrome icon for consistency w/ other Jetpack blocks

jeherve reviewed (approved):
I think this should be good to go for now. I created issues for feedback that was discussed above and other things that I found while testing, but that are not blockers for this PR. I think we should be able to address those issues in follow-up PRs.

I'll merge this now.

https://github.com/Automattic/jetpack/issues/14344
https://github.com/Automattic/jetpack/issues/14346
https://github.com/Automattic/jetpack/issues/14335
https://github.com/Automattic/jetpack/issues/14345
https://github.com/Automattic/jetpack/issues/14343
https://github.com/Automattic/jetpack/issues/14347
